pallets/ajuna-tournament/src/lib.rs

```diff
@@ -0,0 +1,571 @@
+
+	#[pallet::storage]
+	#[pallet::getter(fn rankings)]
+	pub type TournamentRankings<T: Config<I>, I: 'static = ()> = StorageNMap<
+		_,
+		(
+			NMapKey<Identity, T::SeasonId>,
+			NMapKey<Identity, TournamentId>,
+			NMapKey<Identity, T::RankCategory>,
+		),
+		PlayerTableFor<T, I>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn golden_ducks)]
+	pub type GoldenDucks<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+		_,
+		Identity,
+		T::SeasonId,
+		Identity,
+		TournamentId,
+		GoldenDuckStateFor<T, I>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		TournamentCreated { season_id: T::SeasonId, tournament_id: TournamentId },
+		TournamentStarted { season_id: T::SeasonId, tournament_id: TournamentId },
+		TournamentEnded { season_id: T::SeasonId, tournament_id: TournamentId },
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// There's no active tournament for the selected season.
+		NoActiveTournamentForSeason,
+		/// There's already an active tournament for the selected season.
+		AnotherTournamentAlreadyActiveForSeason,
+		/// Cannot find tournament data for the selected season id and tournament id combination.
+		TournamentNotFound,
+		/// Cannot activate a tournament before its configured block start,
+		TournamentActivationTooEarly,
+		/// Cannot deactivate a tournament before its configured block end,
+		TournamentEndingTooEarly,
+		/// An error occurred trying to rank an entity,
+		FailedToRankEntity,
+		/// Tournament configuration is invalid.
+		InvalidTournamentConfig,
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// The account ID of the treasury.
+		pub fn global_treasury_account_id() -> AccountIdFor<T> {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		/// The account ID of the sub-account for a given season_id/tournament_id.
+		pub fn tournament_treasury_account_id(
+			season_id: T::SeasonId,
+			tournament_id: TournamentId,
+		) -> T::AccountId {
+			let account = TournamentTreasuryAccount::<T::SeasonId>::new(
+				T::PalletId::get(),
+				season_id,
+				tournament_id,
+			);
+			account.into_account_truncating()
+		}
+
+		fn ensure_valid_tournament(
+			config: &TournamentConfigFor<T, I>,
+		) -> Result<u16, DispatchError> {
+			ensure!(config.start < config.end, Error::<T, I>::InvalidTournamentConfig);
+
+			if let Some(initial_reward) = config.initial_reward {
+				ensure!(initial_reward > 0_u32.into(), Error::<T, I>::InvalidTournamentConfig);
+			}
+
+			if let Some(max_reward) = config.max_reward {
+				ensure!(max_reward > 0_u32.into(), Error::<T, I>::InvalidTournamentConfig);
+			}
+
+			if let Some(fee_perc) = config.take_fee_percentage {
+				ensure!(fee_perc <= 100, Error::<T, I>::InvalidTournamentConfig);
+			}
+
+			let reward_table_total_dist =
+				config.reward_table.iter().fold(0_u16, |a, b| a + (*b as u16));
```

Comment:
Yeah that's why I cast the inputs to `u16` instead so that a user cannot overflow it, and since the check is basically `reward_table_total_dist <= 100` a user cannot provide a valid config where this would overflow later on payout checks too.

Reply:
Ok, then I think this should be added as a comment. It is very hard to understand this as an uninvolved reader. :)